docs/RFCS/20210330_sql_stats_persistence.md

```diff
@@ -2,36 +2,741 @@
+            "serviceLat": { "$ref": "#/definitions/numeric_stats" },
+            "retryLat":   { "$ref": "#/definitions/numeric_stats" },
+            "commitLat":  { "$ref": "#/definitions/numeric_stats" },
+            "bytesRead":  { "$ref": "#/definitions/numeric_stats" },
+            "rowsRead":   { "$ref": "#/definitions/numeric_stats" }
+          },
+          "required": [
+            "maxRetries",
+            "numRows",
+            "serviceLat",
+            "retryLat",
+            "commitLat",
+            "bytesRead",
+            "rowsRead",
+          ]
+        },
+        "execution_statistics": {
+          "type": "object",
+          "properties": {
+            "cnt":             { "type": "number" },
+            "networkBytes":    { "$ref": "#/definitions/numeric_stats" },
+            "maxMemUsage":     { "$ref": "#/definitions/numeric_stats" },
+            "contentionTime":  { "$ref": "#/definitions/numeric_stats" },
+            "networkMsg":      { "$ref": "#/definitions/numeric_stats" },
+            "maxDiskUsage":    { "$ref": "#/definitions/numeric_stats" },
+          },
+          "required": [
+            "cnt",
+            "networkBytes",
+            "maxMemUsage",
+            "contentionTime",
+            "networkMsg",
+            "maxDiskUsage",
+          ]
+        }
+      },
+
+      "properties": {
+        "stats": { "$ref": "#/definitions/statistics" },
+        "execStats": {
+          "$ref": "#/definitions/execution_statistics"
+        }
+      }
+    }
+    */
+
+    PRIMARY KEY (aggregated_ts, fingerprint_id, app_name, node_id)
+      USING HASH WITH BUCKET_COUNT = 8,
+    INDEX (fingerprint_id, aggregated_ts, app_name, node_id)
+);
+```
+
+The first two columns of the primary keys for both tables contain
+`aggregated_ts` column and `fingerprint_id` column. `aggregated_ts` is the
+timestamp of the beginning of the aggregation interval. This ensures that for
+all entries for every statement and transaction within the same aggregation
+window, the will have the same `aggregated_ts` value. This makes cross-query
+comparison cleaner and also enables us to use `INSERT ON CONFLICT`-style
+statements.
+
+The primary key utilizes hash-sharding with 8 buckets. There are two reasons
+for this design:
+1. Using hash-sharded primary key avoids writing contentions since `aggregated_ts`
+   column contains a monotonically increasing sequence of timestamps. This would
+   allow us to achieve linear scaling.
+1. This speeds up the use case where we want to show aggregated statistics for
+   each fingerprint for the past few hours or days.
+
+The last column in the primary key is `app_name`. This stores the name of the
+application that issued the SQL statement. This is included because same
+statements issued from different applications would have same `fingerprint_id`.
+Therefore, having `app_name` as part of the primary key is important to
```

Comment:
We added database_id to the statement_key recently for db console filtering. Is this already part of the key/fingerprint_id? Do we need to consider this in the design? https://github.com/maryliag/cockroach/commit/f10a7c82e47818ef89bcff8d71a60f261b5fa823